pkg/sql/sqlbase/rowwriter.go

```diff
@@ -842,33 +842,65 @@ func (ru *RowUpdater) UpdateRow(
-		secondaryIndexEntry := secondaryIndexEntries[i]
+	// We're iterating through all of the indexes, which should have corresponding entries in both oldSecondaryIndexEntries
+	// and newSecondaryIndexEntries. Inverted indexes could potentially have more entries at the end of both and we will
+	// update those separately.
+	for i, index := range ru.Helper.Indexes {
+		oldSecondaryIndexEntry := oldSecondaryIndexEntries[i]
+		newSecondaryIndexEntry := newSecondaryIndexEntries[i]
+
+		// We're skipping inverted indexes in this loop, but appending the inverted index entry to the back of
+		// newSecondaryIndexEntries to process later. For inverted indexes we need to remove all old entries before adding
+		// new ones to prevent race conditions.
+		if index.Type == IndexDescriptor_INVERTED {
+			newSecondaryIndexEntries = append(newSecondaryIndexEntries, newSecondaryIndexEntry)
+			oldSecondaryIndexEntries = append(oldSecondaryIndexEntries, oldSecondaryIndexEntry)
+
+			continue
+		}
+
 		var expValue interface{}
-		if !bytes.Equal(newSecondaryIndexEntry.Key, secondaryIndexEntry.Key) {
-			ru.Fks.addCheckForIndex(ru.Helper.Indexes[i].ID)
+		if !bytes.Equal(newSecondaryIndexEntry.Key, oldSecondaryIndexEntry.Key) {
+			ru.Fks.addCheckForIndex(ru.Helper.Indexes[i].ID, ru.Helper.Indexes[i].Type)
 			if traceKV {
-				log.VEventf(ctx, 2, "Del %s", keys.PrettyPrint(ru.Helper.secIndexValDirs[i], secondaryIndexEntry.Key))
+				log.VEventf(ctx, 2, "Del %s", keys.PrettyPrint(ru.Helper.secIndexValDirs[i], oldSecondaryIndexEntry.Key))
 			}
-			batch.Del(secondaryIndexEntry.Key)
-		} else if !bytes.Equal(newSecondaryIndexEntry.Value.RawBytes, secondaryIndexEntry.Value.RawBytes) {
-			expValue = &secondaryIndexEntry.Value
+			batch.Del(oldSecondaryIndexEntry.Key)
+		} else if !bytes.Equal(newSecondaryIndexEntry.Value.RawBytes, oldSecondaryIndexEntry.Value.RawBytes) {
+			expValue = &oldSecondaryIndexEntry.Value
 		} else {
 			continue
 		}
+
 		if traceKV {
 			log.VEventf(ctx, 2, "CPut %s -> %v", keys.PrettyPrint(ru.Helper.secIndexValDirs[i], newSecondaryIndexEntry.Key), newSecondaryIndexEntry.Value.PrettyPrint())
 		}
 		batch.CPut(newSecondaryIndexEntry.Key, &newSecondaryIndexEntry.Value, expValue)
 	}
 
-	// We're deleting indexes in a delete only state.
-	for i, deletedSecondaryIndexEntry := range deleteOldSecondaryIndexEntries {
-		ru.Fks.addCheckForIndex(ru.DeleteHelper.Indexes[i].ID)
-		if traceKV {
-			log.VEventf(ctx, 2, "Del %s", keys.PrettyPrint(ru.DeleteHelper.secIndexValDirs[i], deletedSecondaryIndexEntry.Key))
+	// We're deleting indexes in a delete only state. We're bounding this by the number of indexes because inverted
+	// indexed will be handled separately.
+	if ru.DeleteHelper != nil {
+		for i, deletedSecondaryIndexEntry := range deleteOldSecondaryIndexEntries {
+			if i < len(ru.DeleteHelper.Indexes) {
+				ru.Fks.addCheckForIndex(ru.DeleteHelper.Indexes[i].ID, ru.DeleteHelper.Indexes[i].Type)
```

Comment:
you can totally get rid of the need to check the index in the DELETE_ONLY state. So get rid of this line and thus the if statement above.